src/arizona_template_parser.erl

```diff
@@ -0,0 +1,287 @@
+                   {4, 5},
+                   #{attributes =>
+                      [{<<"charset">>, {text, {4, 19}, <<"UTF-8">>}}],
+                     name => <<"meta">>, is_stateful => false,
+                     is_void => true, inner_content => []}},
+                  {tag,
+                   {5, 5},
+                   #{attributes => [], name => <<"title">>,
+                     is_stateful => false, is_void => false,
+                     inner_content =>
+                      [{expr, {5, 12}, <<"_@title">>}]}},
+                  {tag,
+                   {6, 5},
+                   #{attributes =>
+                      [{<<"src">>,
+                        {text, {6, 17}, <<"assets/js/main.js">>}}],
+                     name => <<"script">>, is_stateful => false,
+                     is_void => false, inner_content => []}}]}},
+             {tag,
+              {8, 1},
+              #{attributes => [], name => <<"body">>,
+                is_stateful => false, is_void => false,
+                inner_content =>
+                 [{tag,
+                   {9, 5},
+                   #{attributes => [], name => <<"h1">>,
+                     is_stateful => false, is_void => false,
+                     inner_content =>
+                      [{text, {9, 9}, <<"Arizona Counter">>}]}},
+                  {block,
+                   {10, 5},
+                   #{attributes =>
+                      [{<<"count">>, {expr, {11, 15}, <<"_@count">>}},
+                       {<<"btn_text">>,
+                        {text, {12, 18}, <<"Increment">>}},
+                       {<<"event">>, {text, {13, 15}, <<"incr">>}}],
+                     name => <<"counter">>}}]}}]}}
+    ]}, parse(tokens_ok())).
+
+parse_unexpected_block_end_test() ->
+    {ok, Tokens} = arizona_template_scanner:scan(~"""
+    <.foo id="bar">
+    """),
+    ?assertEqual({error, {unexpected_block_end, {1, 1}}}, parse(Tokens)).
+
+parse_unexpected_tag_end_test() ->
+    {ok, Tokens} = arizona_template_scanner:scan(~"""
+    <foo id="bar">
+    """),
+    ?assertEqual({error, {unexpected_tag_end, {1, 1}}}, parse(Tokens)).
+
+parse_invalid_directive_test() ->
+    {ok, Tokens} = arizona_template_scanner:scan(~"""
+    <foo :id="bar"></foo>
+    """),
+    ?assertEqual({error, {invalid_directive, {1, 6}}}, parse(Tokens)).
+
+%% --------------------------------------------------------------------
+%% Test support
+%% --------------------------------------------------------------------
+
+tokens_ok() ->
+    {ok, Tokens} = arizona_template_scanner:scan(~"""
+    <!DOCTYPE html>
+    <html lang="en">
+    <head>
+        <meta charset="UTF-8">
+        <title>{_@title}</title>
+        <script src="assets/js/main.js"></script>
+    </head>
+    <body>
+        <h1>Arizona Counter</h1>
+        <.counter
+            count={_@count}
+            btn_text="Increment"
+            event="incr"
+        />
+    </body>
+    </html>
+    """),
+    Tokens.
+
+-endif.
+
```

Comment:
Is this a "real" extra line? Or one of those we discussed? I long for a formatter that is OTP-27 ready 😄

Reply:
One of those we discussed. I'll remove it.

Reply:
[ecf886a](https://github.com/arizona-framework/arizona/pull/151/commits/ecf886a2419f88931c7c6348d8effc8f9e67e93e)

Reply:
[413eb60](https://github.com/arizona-framework/arizona/pull/151/commits/413eb60f3bf6582e22348a723d6d3c8b04539d52)